### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **BUG: Inconsistent HTML tag validation**

The regex patterns for title and description HTML validation are inconsistent:
- Title uses: `/<[^>]*>/g` (rejects any HTML tags)
- Description uses: `/<.*>/g` (allows single `<` or `>` characters)

This inconsistency could lead to security vulnerabilities. Both should use the same strict pattern: `/<[^>]*>/g`

**Example of the issue:**
- `title: "<script>"` → ❌ rejected (good)
- `description: "<script>"` → ✅ accepted (bad)

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
+  if (!id || !/^\d+$/.test(id)) {
+    return res.status(400).json({
+      error: 'Invalid ID format',
+      details: [{ field: 'id', message: 'ID must be a positive integer', value: id }]
+    });
+  }
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:109]: **BUG: ID "0" validation gap**

The current regex `/^\d+$/` allows ID "0", but this might not be valid depending on your ID generation strategy (most auto-increment IDs start from 1).

Consider updating to: `/^[1-9]\d*$/` to only allow positive integers starting from 1, or add explicit documentation about whether ID 0 is valid in your system.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Missing Error Handling**

If `todoStore.create()` throws an exception, the server will crash since there's no try-catch block. Consider adding proper error handling:

```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  console.error('Error creating todo:', error);
  res.status(500).json({ 
    error: 'Internal server error',
    details: [{ field: 'server', message: 'Failed to create todo' }]
  });
}
```

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **Test Coverage Gap**

Missing test case for description HTML tag validation. This should be added to ensure both title and description have consistent HTML protection:

```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      const response = await request(app).get('/abc');
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'id',
+        message: 'ID must be a positive integer',
+        value: 'abc'
+      });
+    });
+
+    // MISSING TEST: Should test ID "0" edge case - currently no validation for this
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:120]: **Test Coverage Gap**

Missing test for ID "0" edge case. Since this is mentioned as a potential bug in the validation logic, it should be tested:

```typescript
it('should reject ID "0" if not valid in system', async () => {
  const response = await request(app).get('/0');
  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'id',
    message: 'ID must be a positive integer',
    value: '0'
  });
});
```

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Review Summary

This PR adds comprehensive input validation to the todo API, which is a **significant security and usability improvement**. The overall approach is solid, but I've identified several critical issues that should be addressed before merging.

### ✅ **Strengths**

1. **Security Enhancement**: HTML tag prevention addresses XSS vulnerabilities
2. **Structured Error Responses**: New error format with field-specific messages is much better for API consumers
3. **Comprehensive Validation**: Length limits, type checking, and sanitization are well implemented
4. **Good Test Coverage**: 15+ test cases cover most scenarios
5. **Clear Documentation**: README updates clearly explain validation rules and breaking changes

### ❌ **Critical Issues**

1. **Security Bug**: Inconsistent HTML validation regex between title and description could allow XSS attacks
2. **Server Stability**: Missing error handling in route handlers could crash the server
3. **ID Validation Gap**: Unclear whether ID "0" should be valid

### 🔧 **Required Changes**

1. **Fix HTML validation consistency** - Use the same strict regex pattern for both title and description
2. **Add error handling** - Wrap `todoStore` operations in try-catch blocks
3. **Clarify ID validation** - Either allow ID "0" explicitly or reject it consistently
4. **Add missing tests** - Cover description HTML validation and ID "0" edge case

### 📋 **Recommendations**

- Consider adding rate limiting validation mentioned in future enhancements
- The `ValidationException` class is defined but never used - consider removing if not needed
- Error messages are clear and user-friendly

This is a **high-quality PR** that significantly improves the codebase. Once the critical issues are resolved, it will be ready to merge.
